File: 2023/06/04/01_kalkulator.py
# Program Python yang meminta pengguna untuk memasukkan dua bilangan bulat dan kemudian menampilkan hasil penjumlahan, pengurangan, perkalian, dan pembagian dari kedua bilangan tersebut.


# metode input() digunakan untuk menerima input dari pengguna
def main():
    while True:
        input_1 = input("Bilangan bulat 1: \n")
        input_2 = input("Bilangan bulat 2: \n")
        try:
            input_1 = int(input_1)
            input_2 = int(input_2)
            if input_1 == 0 or input_2 == 0:
                raise ValueError
            break
        except ValueError:
            print(
                "Bilangan bulat harus berupa angka dan tidak boleh 0. Silakan masukkan kembali."
            )

    # meminta operator matematika dari pengguna
    operators = {
        "+": lambda x, y: x + y,
        "-": lambda x, y: x - y,
        "*": lambda x, y: x * y,
        "/": lambda x, y: x / y,
    }

    # meminta operator matematika dari pengguna
    while True:
        operator = input("Masukkan operator matematika (+, -, *, /): \n")
        if operator in operators:
            hasil = operators[operator](input_1, input_2)
            break
        else:
            print("Operator matematika tidak valid. Silakan masukkan kembali.")

    # menampilkan hasil
    print("Hasil: ", hasil)


if __name__ == "__main__":
    main()

# catatan: lambda adalah fungsi tanpa nama yang dapat mengambil argumen dan mengembalikan nilai. Dalam contoh ini, lambda digunakan untuk mengimplementasikan fungsi penjumlahan, pengurangan, perkalian, dan pembagian.

File: 2023/06/05/01_kalkulator.c
// Program C yang meminta pengguna untuk memasukkan dua bilangan bulat dan kemudian menampilkan hasil penjumlahan, pengurangan, perkalian, dan pembagian dari kedua bilangan tersebut.

#include <stdio.h>

// validasi input
int validasiInput(int input)
{
    while (input == 0)
    {
        printf("Input tidak boleh 0, silahkan masukkan kembali: ");
        scanf("%d", &input);
    }
    return input;
}

// opertor matematika
int tambah(int a, int b)
{
    return a + b;
}

int kurang(int a, int b)
{
    return a - b;
}

int kali(int a, int b)
{
    return a * b;
}

float bagi(int a, int b)
{
    return (float)a / b;
}

// meminta operator dari pengguna
char operator()
{
    char op;
    printf("Masukkan operator (+, -, *, /): ");
    scanf(" %c", &op);
    return op;
}

// menampilkan hasil
void hasil(int a, int b, char op)
{
    switch (op)
    {
    case '+':
        printf("%d + %d = %d\n", a, b, tambah(a, b));
        break;
    case '-':
        printf("%d - %d = %d\n", a, b, kurang(a, b));
        break;
    case '*':
        printf("%d * %d = %d\n", a, b, kali(a, b));
        break;
    case '/':
        printf("%d / %d = %.2f\n", a, b, bagi(a, b));
        break;
    default:
        printf("Operator tidak dikenali\n");
        break;
    }
}

int main()
{
    int a, b;
    char op;

    // meminta input dari pengguna
    printf("Masukkan bilangan pertama: ");
    scanf("%d", &a);

    // meminta operator dari pengguna
    op = operator();

    // meminta input dari pengguna
    printf("Masukkan bilangan kedua: ");
    scanf("%d", &b);

    // validasi input
    a = validasiInput(a);
    b = validasiInput(b);

    // menampilkan hasil
    hasil(a, b, op);

    // memberi opsi untuk mengulang program atau tidak
    char ulang;
    printf("Apakah anda ingin mengulang program? (y/n): ");
    scanf(" %c", &ulang);

    if (ulang == 'y' || ulang == 'Y')
    {
        main();
    }
    else
    {
        printf("Terima kasih telah menggunakan program ini\n");
    }

    return 0;
}